### PR TITLE
Add PCloud support.

### LINKED ISSUE #438: pCloud
Please add pCloud as cloud service to save the db

### LINKED ISSUE #512: feature request: support for pcloud
Please add support for pcloud. Pcloud is the only cloudspace with proper Linux support after dropbox ditches support for encrpted ext4 fs in November.

### PR DESCRIPTION
**Pull request by gilbsgilbs**
Fixes #512 
Fixes #438 

This PR adds PCloud integration to Keepass2Android. Here are some notes:

- Not sure what you want to do with the OAuth 2 client id. If you want to take the ownership on it, please feel free and I'll remove my app from my PCloud account. Otherwise, if you trust me enough, I can let the app untouched in my personal PCloud account.
- I did not implement the global account access. I find the practice unsafe and on the top of that, given some limitations of their SDK, it would have been quite heavy to implement (code-wise and network-wise) as I see it. Consequently, unlike Dropbox or OneCloud, KP2A has only access to its very own folder on PCloud and I think it's fine.
- I had to create a `PCloudBindings` project because of [this bug](https://bugzilla.xamarin.com/show_bug.cgi?id=44573). PCloud SDK provides an OAuth 2 helper library shipped as an Android AAR artifact. I couldn't import the AAR in `JavaFileStorageBindings` because it already has an AAR in it and it wouldn't merge resources properly.
- I got the JAR and AAR files from [bintray](https://bintray.com/pcloud/pcloud-sdk/) if you want to perform some checksums.
- Since Android WebView doesn't handle custom schemes, I was not able to have a proper landing page after the OAuth 2 redirection. Therefore, at the end of the authentication on PCloud, you can briefly see an error landing page. While that's unfortunate, I don't see an easy solution for this, apart from creating our own "success" webpage or something like this. Maybe we could also open an issue on the PCloud SDK issue tracker because only them can close the WebView earlier (or maybe they can't, I'm not sure).
- Tested in a VM and on my personal OnePlus 5.
- Since I was not able to find a proper tooling to work with on Xamarin.Android project on Linux, I did most of the C# work using vim (I know, it sounds insane huh). Although I did my best, I might have forgotten to commit some files. If you notice anything missing, it might be just me so please let me know before digging further.

Thanks.

### CODE REVIEW

**Comment by PhilippC** [comment]
@gilbsgilbs thanks a lot for the PR! I haven't tested it, but it looks pretty good! I have just one comment.

File: src/java/JavaFileStorage/app/src/main/java/keepass2android/javafilestorage/PCloudFileStorage.java [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+    public boolean checkForFileChangeFast(String path, String previousFileVersion) throws Exception {
+        if (previousFileVersion == null || "".equals(previousFileVersion)) {
+            return false;
+        }
+
+        path = this.cleanPath(path);
+
+        RemoteFile remoteFile = this.getRemoteFileByPath(path);
+
+        return !remoteFile.hash().equals(previousFileVersion);
```
[the remark this answers is not in the document]
Comment (gilbsgilbs) [reply src/java/JavaFileStorage/app/src/main/java/keepass2android/javafilestorage/PCloudFileStorage.java]: It's [received from PCloud API](https://docs.pcloud.com/methods/folder/listfolder.html). It's think it's casted to a string by introspection in GSON. (see https://github.com/pCloud/pcloud-sdk-java/blob/master/java-core/src/main/java/com/pcloud/sdk/internal/RealRemoteFile.java#L43).

**Comment by gilbsgilbs** [comment]
@PhilippC Many thanks for the quick review :) .

**Comment by gilbsgilbs** [comment]
Thanks for your time @PhilippC.

**Comment by PhilippC** [comment]
thanks for yours, @gilbsgilbs!
I just built your code and it is working! Just two things that may or may not be improved, depending on the SDK:
* Is it possible to check if there is an auth token in the constructor and if yes, return false from requiresSetup()? That allows to skip the "FileSelectActivity" and proceed to PasswordActivity directory. (It is something that is freqently criticised by Google Drive users)
* The current implementation assumes the user only has one account ot PCloud. While I think that's a valid assumption for most users, I have learnt that with a large user base there will be quite a few users for which the assumption won't hold. E.g. it's the same with my OneDrive implementation but some users have requested to allow switching of accounts when clicking "Open database". If it's easily possible to do so, that would be nice, but it's not a big thing, IMHO.

**Comment by gilbsgilbs** [comment]
@PhilippC glad it worked.

> Is it possible to check if there is an auth token in the constructor and if yes, return false from requiresSetup()? That allows to skip the "FileSelectActivity" and proceed to PasswordActivity directory. (It is something that is freqently criticised by Google Drive users)

Not sure I understand all the implication of this. Is it okay to make `requiresSetup` reflect the behavior from `isConnected()`? How can I test it works?

```java
    @Override
    public boolean requiresSetup(String path) {
        return !this.isConnected();
    }
```

> The current implementation assumes the user only has one account ot PCloud. While I think that's a valid assumption for most users, I have learnt that with a large user base there will be quite a few users for which the assumption won't hold. E.g. it's the same with my OneDrive implementation but some users have requested to allow switching of accounts when clicking "Open database". If it's easily possible to do so, that would be nice, but it's not a big thing, IMHO.

This one would either require to submit a PR to the SDK to allow clearing browsing data before showing the WebView or re-implement all the OAuth dance on our end (either natively or using some library I'm unaware of). Honestly, I'm not sure I want to spend so much effort into fixing such a specific use-case at the moment. I could re-consider if KP2A gets a lot of PCloud users which I don't think.

I'm more concerned by the fact that the current implementation cannot read/write everywhere. On the security perspective it's definitely the best option, yet I can see legitimate and probably quite frequent workflows where this is a painpoint. For example if you have another client on another device that has the same limitation, you'll not be able to synchronize. But given the current SDK, you can't just request the file `x/y/z.kdbx` directly, you have to iterate through all the tree manually (unless I'm missing something). So either you make a lot of network calls (and `getCurrentFileVersionFast` gets not so fast anymore), or you request the whole tree recursively (as done in the current implementation) which is simpler but can be dangerous (and consume even more network if you have a lot of files). Anyways, if there's no solution, there's no problem, right? Still, this might be something to keep in mind or tell to PCloud guys.

**Comment by gilbsgilbs** [comment]
Sorry for the delay. I'm still waiting for them to fix a bug with tokens revocation :man_facepalming: . Because of this, I can't test `requiresSetup` properly at the moment. Although they told me they were aware of the bug, they did not give me any ETA.

**Comment by gilbsgilbs** [comment]
@PhilippC, you can mention gilbsgilbs. Thanks for the release.

**Comment by PhilippC** [comment]
ok, will do so starting from 1.07-pre2. (pre1 was already built and is now in beta!)

**Pull request merged**